### DISCUSSION TITLE
RAW RGB file to BMP

### DISCUSSION THREAD

**Discussion by avironman** [ImageMagick/ImageMagick#5317] [Help] [answered]
Hello All,
I'm trying to debayer a raw file containing 8bit RGB image 720x540 pixels into a png or bmp file output, but I get an unexpected end-of-file error no matter what I try...
I checked that the file size is equal to width*height, but no way to make it work.

Here is the command `magick -verbose -size 720x540 -depth 8 rgb:"bytearray.raw" test.bmp`

When I try `magick -verbose -size 720x540 -depth 8 gray:"bytearray.raw" test.bmp` I get a result but not what I expect

Please find attached the raw data along with the bmp decoded using a different software.
[RAW.zip](https://github.com/ImageMagick/ImageMagick/files/9086956/RAW.zip)
.
Any help would be appreciated.

**Comment by snibgo** [comment]
I am totally confused. The file has 388800 bytes. How many pixels do you think it has? How many channels? How many bits per channel per pixel?

For example, it might have 720x540 pixels, each with only one channel (so it is "gray:", not "rgb:"), 8 bits/channel/pixel.
```
magick -verbose -size 720x540 -depth 8 gray:"bytearray.raw" x.png
```
The output x.png looks reasonable, and like a Bayer image.

**Reply by avironman** [reply]
Hello,
File has 720x540 pixels so it is 388800 pixels in total. Each pixel is 8bit, organized as follow Rg-GB.

It comes from a machine vision camera, so it's trully raw, each byte is the real pixel of the CMOS sensor elements.

I want to demosaic this image, into a 24 or 32 bits bmp color format. How would you proceed?

**Reply by avironman** [reply]
OK I see you're right, I posted wrong line of code ;), my mistake, but even giving the right width and heigth it does not work

**Reply by avironman** [reply]
I think IM is not able to do this I will check if dcraw is an option, thanks anyway for your time

**Comment by snibgo** [comment, marked as answer]
It is a Bayer image. This means each pixel has only one channel, which represents either red or green or blue.

IM has no problem reading the file, with the command I showed.

IM does not contain de-Bayering code. But the process can be regarded as simple interpolation, which IM can do. Alternatively, you can use IM and exiftool to make a DNG file, which any raw processor (such as dcraw or darktable or many others) can read and process. See [Processing Bayer pixels](http://im.snibgo.com/procbay.htm)

**Reply by avironman** [reply]
OK Thanks for your reply, I thought using RGB: it would tell IM that it was a bayer image, my mistake.
Thanks for your support, I'm already looking at your suggestion.

**Comment by snibgo** [comment]
Another link for you: [Demosaicing](http://im.snibgo.com/demosaic.htm). This shows how we can demosaic using just ImageMagick.
```
%IMG7%magick -verbose -size 720x540 -depth 8 gray:"bytearray.raw" x.png

call %PICTBAT%demos x.png
```
The output is x_dm.png:
![x_dm](https://user-images.githubusercontent.com/33812368/178501957-3a5b0da7-4884-4e70-93e1-5fcf336d5f3f.png)

**Comment by urban-warrior** [comment]
We added support for reading raw Bayer (rggb) images based on http://im.snibgo.com/demosaic.htm.  The command for your image is:

```
magick -verbose -size 720x540 -depth 8 bayer:bytearray.raw  bottle.png
```

The RMSE between your results and ImageMagick's is 0.0237335.

The support is in ImageMagick beta, available in less than an hour.  We're working on support for writing Bayer images.  Look for it in a day or two.
